File: tourist_landmark/bfs.py
from collections import deque

class Graph:
    def __init__(self):
        # adjacency list representation of the graph
        self.adj_list = {}

    def add_edge(self, u, v):
        """Adds an undirected edge between u and v"""
        if u not in self.adj_list:
            self.adj_list[u] = []
        if v not in self.adj_list:
            self.adj_list[v] = []

        # add both ways since the graph is undirected
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)

    def bfs_all_paths(self, start, goal):
        """
        Finds all shortest paths from start to goal using BFS.
        Returns a list of shortest paths (each path is a list of nodes),
        or None if no path exists.
        """
        # check if start and goal exist
        if start not in self.adj_list or goal not in self.adj_list:
            return None

        # queue of paths to explore
        frontier = deque([[start]])

        # set of explored nodes
        explored = set()

        # stores all shortest paths
        shortest_paths = []
        found_level = None  # shortest path length (once found)

        while frontier:
            # get the shallowest path
            path = frontier.popleft()
            node = path[-1]

            # check if goal is reached
            if node == goal:
                if found_level is None:
                    found_level = len(path)

                if len(path) == found_level:
                    shortest_paths.append(path)

                # don’t expand paths beyond the goal
                continue

            explored.add(node)

            # explore neighbors
            for neighbor in self.adj_list.get(node, []):
                if neighbor not in explored and neighbor not in path:
                    new_path = path + [neighbor]
                    frontier.append(new_path)

        return shortest_paths if shortest_paths else None

File: tourist_landmark/bfs.js
// Define a Graph class using an adjacency list representation
class Graph {
  constructor() {
    // Initialize an empty adjacency list
    this.adjList = {};
  }

  // Adds an undirected edge between node u and node v
  addEdge(u, v) {
    // Create adjacency list entries if they don't exist
    if (!this.adjList[u]) this.adjList[u] = [];
    if (!this.adjList[v]) this.adjList[v] = [];

    // Add each node to the other's adjacency list
    this.adjList[u].push(v);
    this.adjList[v].push(u);
  }

  /**
   * Finds all shortest paths from `start` node to `goal` node using BFS.
   * Returns an array of all shortest paths (each path is an array of nodes),
   * or null if no path exists.
   */
  bfsAllPaths(start, goal) {
    // If either the start or goal node doesn't exist in the graph, return null
    if (!this.adjList[start] || !this.adjList[goal]) return null;

    // Frontier stores paths to be explored; starts with a single path containing the start node
    let frontier = [[start]];

    // Set to keep track of fully explored nodes (to avoid cycles)
    let explored = new Set();

    // Array to store all shortest paths found
    let shortestPaths = [];

    // Once the shortest path length is known, all paths must match it
    let foundLevel = null;

    // Loop until there are no more paths to explore
    while (frontier.length > 0) {
      // Remove the shallowest path (FIFO behavior of a queue)
      let path = frontier.shift();

      // Get the last node in the current path
      let node = path[path.length - 1];

      // Check if the goal has been reached
      if (node === goal) {
        // If this is the first time reaching the goal, record the path length
        if (foundLevel === null) {
          foundLevel = path.length;
        }

        // If current path matches the shortest length, add to result
        if (path.length === foundLevel) {
          shortestPaths.push(path);
        }

        // Skip further processing for this path (we don’t extend goal paths)
        continue;
      }

      // Mark the current node as explored
      explored.add(node);

      // Explore all neighbors of the current node
      for (let neighbor of this.adjList[node] || []) {
        // Skip neighbor if it has already been explored
        // Also avoid cycles by not revisiting nodes already in the current path
        if (!explored.has(neighbor) && !path.includes(neighbor)) {
          // Create a new path extended by the neighbor and enqueue it
          frontier.push([...path, neighbor]);
        }
      }
    }

    // If at least one shortest path was found, return them; otherwise, return null
    return shortestPaths.length > 0 ? shortestPaths : null;
  }
}
